pkg/resources/kafka/headlessService.go

```diff
@@ -92,3 +97,95 @@ func (r *Reconciler) deleteHeadlessService() error {
 
 	return err
 }
+
+// deleteNonHeadlessServices deletes the all-broker service that was created for the current KafkaCluster
+// if there is any and also the service of each broker
+func (r *Reconciler) deleteNonHeadlessServices(ctx context.Context) error {
+	svc := corev1.Service{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: corev1.SchemeGroupVersion.String(),
+			Kind:       "Service",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: r.KafkaCluster.GetNamespace(),
+			Name:      fmt.Sprintf(kafkautils.AllBrokerServiceTemplate, r.KafkaCluster.GetName()),
+		},
+	}
+
+	err := r.Client.Delete(ctx, &svc)
+	if err != nil && client.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	// delete broker services
+	labelSelector := labels.NewSelector()
+	for k, v := range apiutil.LabelsForKafka(r.KafkaCluster.GetName()) {
+		req, err := labels.NewRequirement(k, selection.Equals, []string{v})
+		if err != nil {
+			return err
+		}
+		labelSelector = labelSelector.Add(*req)
+	}
+
+	// add "has label 'brokerId' to matching labels selector expression
+	req, err := labels.NewRequirement(v1beta1.BrokerIdLabelKey, selection.Exists, nil)
+	if err != nil {
+		return err
+	}
+	labelSelector = labelSelector.Add(*req)
+
+	var services corev1.ServiceList
+	err = r.Client.List(ctx, &services,
+		client.InNamespace(r.KafkaCluster.GetNamespace()),
+		client.MatchingLabelsSelector{Selector: labelSelector},
+	)
+
+	if err != nil {
+		return errors.WrapIfWithDetails(err, "failed to list services",
+			"namespace", r.KafkaCluster.GetNamespace(),
+			"label selector", labelSelector.String())
+	}
+
+	// if NodePort is used for any of the external listeners, the corresponding services need to remain
+	// so that clients from outside the Kubernetes cluster can reach the brokers
+	filteredSvcsToDelete := services
+	if isNodePortAccessMethodInUseAmongExternalListeners(r.KafkaCluster.Spec.ListenersConfig.ExternalListeners) {
+		filteredSvcsToDelete = nonNodePortServices(services)
+	}
+
+	for i := range filteredSvcsToDelete.Items {
```

Comment:
Here we can use _, svc = range filteredSvcsToDelete.Items

Reply:
right... I tried not to touch the implementation that was copied over. But updated in commit a4a8fb53ccba2bb8dd6be6d2c809ca778f005bb0